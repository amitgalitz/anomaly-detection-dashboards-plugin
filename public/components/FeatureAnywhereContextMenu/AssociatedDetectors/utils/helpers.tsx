/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';
import { EuiBasicTableColumn, EuiHealth, EuiLink } from '@elastic/eui';
import { DETECTOR_STATE } from 'server/utils/constants';
import { stateToColorMap } from '../../../../pages/utils/constants';
import { PLUGIN_NAME } from '../../../../utils/constants';
import { Detector } from '../../../../models/interfaces';

export const renderState = (state: DETECTOR_STATE) => {
  return (
    //@ts-ignore
    <EuiHealth color={stateToColorMap.get(state)}>{state}</EuiHealth>
  );
};

export const getColumns = ({ handleUnlinkDetectorAction }) =>
  [
    {
      field: 'name',
      name: 'Detector',
      sortable: true,
      truncateText: true,
      width: '30%',
      align: 'left',
      render: (name: string, detector: Detector) => (
        <EuiLink href={`${PLUGIN_NAME}#/detectors/${detector.id}`}>
          {name}
        </EuiLink>
      ),
    },
    {
      field: 'curState',
      name: 'Real-time state',
      sortable: true,
      align: 'left',
      width: '30%',
      truncateText: true,
      render: renderState,
    },
    {
      field: 'totalAnomalies',
      name: 'Anomalies/24hr',
      sortable: true,
      dataType: 'number',
      align: 'left',
      truncateText: true,
      width: '30%',
    },
    {
      name: 'Actions',
      align: 'left',
      truncateText: true,
      width: '10%',
      actions: [
        {
          type: 'icon',
          name: 'Unlink Detector',
          description: 'Unlink Detector',
          icon: 'unlink',
          onClick: handleUnlinkDetectorAction,
        },
      ],
    },
  ] as EuiBasicTableColumn<any>[];

export const search = {
  box: {
    incremental: true,
    schema: true,
  },
};
