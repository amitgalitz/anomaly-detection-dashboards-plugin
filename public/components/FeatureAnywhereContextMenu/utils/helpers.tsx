//   /*
// * Copyright OpenSearch Contributors
// * SPDX-License-Identifier: Apache-2.0
// */

// import { useState, useEffect } from 'react';
// import _ from 'lodash';
// import { getSavedFeatureAnywhereLoader } from '../../../services';
// import { SavedObjectLoader } from '../../../../../../src/plugins/saved_objects/public';
// import { ISavedAugmentVis } from '../../../../../../src/plugins/vis_augmenter/public';

// import { get, isEmpty } from 'lodash';

// export const getAllAssociatedDetectors = (embeddable) => {
//     const savedObjectLoader: SavedObjectLoader = getSavedFeatureAnywhereLoader();
//     savedObjectLoader.findAll().then((resp: any) => {
//         if (resp != undefined) {
//           const savedAugmentObjectsArr: ISavedAugmentVis[] = get(
//             resp,
//             'hits',
//             []
//           );
//           const currentAssociatedDetectors = getAssociatedDetectors(
//             Object.values(allDetectors),
//             savedAugmentObjectsArr
//           );
//           console.log("currentAssociatedDetectors: " + currentAssociatedDetectors)
//           const curDetectorsToDisplayOnList = getExistingDetectorsAvalableToAssociate(currentAssociatedDetectors)
//           setExistingDetectorsAvailableToAssociate(curDetectorsToDisplayOnList);
//           setIsLoadingFinalDetectors(false);
//         }
//       });
// }

//   // Handle all changes in the assoicated detectors such as unlinking or new detectors associated
//   useEffect(() => {
//     // Gets all augmented saved objects

//   }, [allDetectors]);

//   const getExistingDetectorsAvalableToAssociate = (currentAssociatedDetectors: DetectorListItem[]) => {
//     console.log("current associated detectors: " + JSON.stringify(currentAssociatedDetectors))
//     return currentAssociatedDetectors

//   }

//   // cross checks all the detectors that exist with all the savedAugment Objects to only display ones
//   // that are associated to the current visualization
//   const getAssociatedDetectors = (
//     detectors: DetectorListItem[],
//     savedAugmentObjects: ISavedAugmentVis[]
//   ) => {
//     // Filter all savedAugmentObjects that aren't linked to the specific visualization
//     console.log("vizID: " + JSON.stringify(embeddableVisId));
//     const savedAugmentForThisVisualization: ISavedAugmentVis[] =
//       savedAugmentObjects.filter(
//         (savedObj) => get(savedObj, 'visId', '') === embeddableVisId
//       );

//       console.log("savedAugmentForThisVisualization: " + JSON.stringify(savedAugmentForThisVisualization))

//     // Map all detector IDs for all the found augmented vis objects
//     const savedAugmentDetectorsSet = new Set(
//       savedAugmentForThisVisualization.map((savedObject) =>
//         get(savedObject, 'pluginResourceId', '')
//       )
//     );
//     console.log("savedAugmentDetectorsSet: " + JSON.stringify(savedAugmentDetectorsSet))

//     // filter out any detectors that aren't on the set of detectors IDs from the augmented vis objects.
//     const detectorsToDisplay = detectors.filter((detector) =>
//       savedAugmentDetectorsSet.has(detector.id)
//     );
//     return detectorsToDisplay;
//   };
