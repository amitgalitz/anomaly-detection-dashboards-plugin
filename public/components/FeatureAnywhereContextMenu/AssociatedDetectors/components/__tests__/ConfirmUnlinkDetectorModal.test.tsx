/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { findAllByTestId, render, waitFor } from '@testing-library/react';
import { ConfirmUnlinkDetectorModal } from '../index';
import { getRandomDetector } from '../../../../../../public/redux/reducers/__tests__/utils';
import { DetectorListItem } from '../../../../../../public/models/interfaces';
import userEvent from '@testing-library/user-event';

describe('ConfirmUnlinkDetectorModal spec', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const testDetectors = [
    {
      id: 'detectorId1',
      name: 'test-detector-1',
    },
    {
      id: 'detectorId2',
      name: 'test-detector-2',
    },
  ] as DetectorListItem[];

  const ConfirmUnlinkDetectorModalProps = {
    detector: testDetectors[0],
    onHide: jest.fn(),
    onConfirm: jest.fn(),
    onUnlinkDetector: jest.fn(),
    isListLoading: false,
  };

  test('renders the component correctly', () => {
    const { container, getByText } = render(
      <ConfirmUnlinkDetectorModal {...ConfirmUnlinkDetectorModalProps} />
    );
    getByText('Remove association?');
    getByText(
      'Removing association unlinks test-detector-1 detector from the visualization but does not delete it. The detector association can be restored.'
    );
    expect(container).toMatchSnapshot();
  });
  test('should call onConfirm() when closing', async () => {
    const { container, getByText, getByTestId } = render(
      <ConfirmUnlinkDetectorModal {...ConfirmUnlinkDetectorModalProps} />
    );
    getByText('Remove association?');
    await waitFor(() => {});
    userEvent.click(getByTestId('confirmUnlinkButton'));
    await waitFor(() => {});
    expect(ConfirmUnlinkDetectorModalProps.onConfirm).toHaveBeenCalled();
  });
  test('should call onConfirm() when closing', async () => {
    const { container, getByText, getByTestId } = render(
      <ConfirmUnlinkDetectorModal {...ConfirmUnlinkDetectorModalProps} />
    );
    getByText('Remove association?');
    await waitFor(() => {});
    userEvent.click(getByTestId('confirmUnlinkButton'));
    await waitFor(() => {});
    expect(ConfirmUnlinkDetectorModalProps.onConfirm).toHaveBeenCalled();
  });
  test('should call onHide() when closing', async () => {
    const { getByTestId } = render(
      <ConfirmUnlinkDetectorModal {...ConfirmUnlinkDetectorModalProps} />
    );
    await waitFor(() => {});
    userEvent.click(getByTestId('cancelUnlinkButton'));
    await waitFor(() => {});
    expect(ConfirmUnlinkDetectorModalProps.onHide).toHaveBeenCalled();
  });
});
