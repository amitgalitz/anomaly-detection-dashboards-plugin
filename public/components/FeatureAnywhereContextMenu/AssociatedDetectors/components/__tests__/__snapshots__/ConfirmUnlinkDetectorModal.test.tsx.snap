// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeModal spec renders the component 1`] = `
<div
  aria-hidden="true"
  data-aria-hidden="true"
/>
`;

exports[`ConfirmUnlinkDetectorModal spec renders the component 1`] = `
<div
  aria-hidden="true"
  data-aria-hidden="true"
/>
`;

exports[`ConfirmUnlinkDetectorModal spec renders the component correctly 1`] = `
<div
  aria-hidden="true"
  data-aria-hidden="true"
/>
`;
